torch/csrc/jit/interpreter.cpp

```diff
@@ -0,0 +1,288 @@
+using Callback = std::function<void(InputList, OutputList)>;
+// Returns a function implementing functionality of a given node,
+// or nullptr if it's a no-op for autograd.
+Callback getCallback(Node *node) {
+  IR_IFM(node, PythonOp)
+    throw NotImplementedException();
+  IR_ELSEIFM(CppOp)
+    throw NotImplementedException();
+  IR_ELSEIF(Select)
+    barf("getCallback() on select?");
+  IR_ELSEIF(FusionGroup)
+#ifdef WITH_CUDA
+    auto fusion_fn = sharedFusionCompiler().getOrCompile(*value->g(kSubgraph));
+    return [fusion_fn](InputList inputs, OutputList outputs) {
+      fusion_fn->launch(inputs, outputs);
+    };
+#else
+    throw std::runtime_error("don't know how to execute FusionGroups without CUDA");
+#endif
+  IR_ELSEIF(Constant)
+    auto t = value->t(kvalue);
+    return [t](InputList inputs, OutputList outputs) {
+      outputs.push_back(t);
+    };
+  IR_ELSEIF(Undefined)
+    return [](InputList inputs, OutputList outputs) {
+      outputs.push_back(at::Tensor());
+    };
+  IR_ELSE()
+    return getTensorOp(node).op;
+  IR_END()
+}
+
+
+// We need some lists for inputs and outputs. To keep all the memory
+// contiguous we allocate a single vector and use offsets into the vector
+// which are stored in the RegList struct
+// start is an offset into int_data of Function if this list is integers
+// and bool_data of Function if this list is booleans (only free_flags)
+struct RegList {
+  int start;
+  int size;
+};
+
+struct UseList {
+  // values to be used
+  RegList values;
+  // boolean flags indicating whether to free the Tensor after this use
+  RegList free_flags;
+};
+
+// one instruction plus meta-data
+struct Instruction {
+  Callback callback;
+  UseList inputs;
+  RegList outputs;
+};
+
+struct Stage {
+  RegList inputs; // inputs to define for the stage
+  UseList outputs; // values consumed by the return
+  std::vector<Instruction> instructions;
+};
+
+// pre-processing that happens once per graph
+struct FunctionImpl {
+  FunctionImpl(std::shared_ptr<Graph> & graph)
+  : graph(graph) {
+    int64_t cur_stage = -1;
+    size_t input_pos = 0;
+    size_t output_pos = 0;
+    // step 1: encode all operators and stages into registers and fill in
+    // input/output lists
+    for(auto node : graph->nodes()) {
+      if(node->kind() == kSelect)
+        continue;
+      insertStagesTo(cur_stage, node->stage(), input_pos, output_pos);
+      cur_stage = node->stage();
+      stages.back().instructions.emplace_back();
+      auto & inst = stages.back().instructions.back();
+      intListBegin(inst.inputs.values);
+      for(auto input : node->inputs()) {
+        intListInsert(inst.inputs.values, getOrAllocateRegister(input, true));
+      }
+      intListBegin(inst.outputs);
+      for(auto output : node->outputs()) {
+        intListInsert(inst.outputs, getOrAllocateRegister(output));
+      }
+      inst.callback = getCallback(node);
+    }
+    insertStagesTo(cur_stage, graph->stage(), input_pos, output_pos);
```

Comment:
Why is this necessary?

Reply:
It is possible (though unlikely), that a final stage is just an identity function of its inputs so there won't be instructions. So this makes sure we have inserted Stage objects for each stage.

Reply:
Can you please add a comment? It's not immediately obvious